packages/lib-runtime/src/store/PluginStore.ts

```diff
@@ -183,53 +110,76 @@ export class PluginStore<TLoadedExtension extends LoadedExtension = LoadedExtens
     };
   }
 
-  private invokeListeners(eventTypes: PluginStoreEventType[]) {
+  private invokeListeners(eventTypes: PluginEventType[]) {
     eventTypes.forEach((t) => {
       this.listeners.get(t)?.forEach((listener) => {
         listener();
       });
     });
   }
 
-  /**
-   * Add new plugin to the `PluginStore`.
-   *
-   * Once added, the plugin is disabled by default. Enable it to put its extensions into use.
-   *
-   * Returns `true` if the plugin was added successfully.
-   */
-  addPlugin(metadata: PluginMetadata, processedExtensions: TLoadedExtension[]) {
-    const pluginName = metadata.name;
-    const pluginVersion = metadata.version;
+  getExtensions() {
+    return [...this.extensions];
+  }
 
-    if (this.loadedPlugins.has(pluginName)) {
-      consoleLogger.warn(`Attempt to re-add an already loaded plugin ${pluginName}`);
-      return false;
-    }
+  getPluginInfo() {
+    const loadedEntries = Array.from(this.loadedPlugins.keys()).reduce((acc, pluginName) => {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const plugin = this.loadedPlugins.get(pluginName)!;
 
-    this.loadedPlugins.set(pluginName, {
-      metadata: Object.freeze(metadata),
-      extensions: processedExtensions.map((e) => Object.freeze(e)),
-      enabled: false,
-    });
+      acc.push({
+        pluginName,
+        status: 'loaded',
+        metadata: plugin.metadata,
+        enabled: plugin.enabled,
+      });
 
-    this.failedPlugins.delete(pluginName);
-    this.invokeListeners([PluginStoreEventType.PluginLoaded]);
+      return acc;
+    }, [] as PluginInfoEntry[]);
 
-    consoleLogger.info(`Added plugin ${pluginName} version ${pluginVersion}`);
+    const failedEntries = Array.from(this.failedPlugins.values()).reduce((acc, pluginName) => {
+      acc.push({ pluginName, status: 'failed' });
+      return acc;
+    }, [] as PluginInfoEntry[]);
 
-    return true;
+    const pendingEntries = (this.loader?.getPendingPluginNames() ?? []).reduce(
+      (acc, pluginName) => {
+        acc.push({ pluginName, status: 'pending' });
+        return acc;
+      },
+      [] as PluginInfoEntry[],
+    );
+
+    return [...loadedEntries, ...failedEntries, ...pendingEntries];
+  }
+
+  async loadPlugin(baseURL: string) {
+    if (this.loader === undefined) {
```

Comment:
You should be checking for null values instead of explicitly undefined.
```suggestion
    if (this.loader == null) {
```
or
```suggestion
    if (!this.loader) {
```

Reply:
`this.loader` is marked as private and its value is set via `setLoader` method so this check should apply there, I think.

I don't like `!falsy` expressions too much because [truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) encompasses a lot of different value types.

I'll update this to `instanceof` check since `this.loader` must be an object created via `PluginLoader` constructor function.

Reply:
Ah `loader` has type `| undefined`. This is ok then. Sorry I missed that.
No need for `instanceof` check since we have typescript to handle that for us.